# [Introduction to Data Science](http://datascience-intro.github.io/1MS041-2024/)    
## 1MS041, 2024 
&copy;2024 Raazesh Sainudiin, Benny Avelin. [Attribution 4.0 International     (CC BY 4.0)](https://creativecommons.org/licenses/by/4.0/)

# 01. BASH Unix Shell 


1. Dropping into BASH (Unix Shell) and using basic Shell commands
    * `pwd` --- print working directory
    * `ls` --- list files in current working directory
    * `mkdir` --- make directory
    * `cd` --- change directory
    * `man ls` --- manual pages for any command
    * `head` --- show the first lines of a file
2. Grabbing files from the internet using `curl`

In [1]:
def showURL(url, ht=500):
    """Return an IFrame of the url to show in notebook with height ht"""
    from IPython.display import IFrame
    return IFrame(url, width='95%', height=ht)


showURL('https://en.wikipedia.org/wiki/Bash_(Unix_shell)', 400)

## 1. Dropping into BASH (Unix Shell)

Using `%%sh` in a code cell we can access the BASH (Unix Shell) command prompt.

Let us `pwd` or print working directory.

In [2]:
%%sh
pwd

/Users/madhurgupta/Documents/HT2024/Introduction to Data Science/Code


In [3]:
%%sh
# this is a comment in BASH shell as it is preceeded by '#'
ls # list the contents of this working directory

01-Probability.ipynb
A01-BASH_Unix_Shell.ipynb
Assignment_1.ipynb
Utils.py
data
images
main.py


In [4]:
%%sh
mkdir -p mydir

In [5]:
%%sh
cd mydir
pwd
ls -al

/Users/madhurgupta/Documents/HT2024/Introduction to Data Science/Code/mydir
total 0
drwxr-xr-x@  2 madhurgupta  staff   64 Sep  5 08:42 .
drwxr-xr-x@ 14 madhurgupta  staff  448 Sep  5 08:42 ..


In [6]:
%%sh
pwd

/Users/madhurgupta/Documents/HT2024/Introduction to Data Science/Code


## "Use the source" by `man`-ning the unknown `command`

By evaluating the next cell, you are using the `man`ual pages to find more about the command `ls`. You can learn more about any command called `command` by typing `man command` in the BASH shell.

The output of the next cell with command `man ls` will look something like the following:

```
LS(1)                            User Commands                           LS(1)

NAME
       ls - list directory contents

SYNOPSIS
       ls [OPTION]... [FILE]...

DESCRIPTION
       List  information  about  the FILEs (the current directory by default).
       Sort entries alphabetically if none of -cftuvSUX nor --sort  is  speci‐
       fied.

       Mandatory  arguments  to  long  options are mandatory for short options
       too.

       -a, --all
              do not ignore entries starting with .

       -A, --almost-all
              do not list implied . and ..
...
...
...
   Exit status:
       0      if OK,

       1      if minor problems (e.g., cannot access subdirectory),

       2      if serious trouble (e.g., cannot access command-line argument).

AUTHOR
       Written by Richard M. Stallman and David MacKenzie.

REPORTING BUGS
       GNU coreutils online help: <http://www.gnu.org/software/coreutils/>
       Report ls translation bugs to <http://translationproject.org/team/>

COPYRIGHT
       Copyright © 2017 Free Software Foundation, Inc.   License  GPLv3+:  GNU
       GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
       This  is  free  software:  you  are free to change and redistribute it.
       There is NO WARRANTY, to the extent permitted by law.

SEE ALSO
       Full documentation at: <http://www.gnu.org/software/coreutils/ls>
       or available locally via: info '(coreutils) ls invocation'

GNU coreutils 8.28               January 2018                            LS(1)
```

In [7]:
%%sh
## uncomment by removing '#' in the next line and try executing this cell
man ls

LS(1)                       General Commands Manual                      LS(1)

NAME
     ls – list directory contents

SYNOPSIS
     ls [-@ABCFGHILOPRSTUWabcdefghiklmnopqrstuvwxy1%,] [--color=when]
        [-D format] [file ...]

DESCRIPTION
     For each operand that names a file of a type other than directory, ls
     displays its name as well as any requested, associated information.  For
     each operand that names a file of type directory, ls displays the names
     of files contained within that directory, as well as any requested,
     associated information.

     If no operands are given, the contents of the current directory are
     displayed.  If more than one operand is given, non-directory operands are
     displayed first; directory and non-directory operands are sorted
     separately and in lexicographical order.

     The following options are available:

     -@      Display extended attribute keys and sizes in long (-l) output.

     -A      Include directory entr

## 2. Grabbing files from internet using curl

In [8]:
%%sh
cd mydir
curl -O http://lamastex.org/datasets/public/SOU/sou/20170228.txt

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 29323  100 29323    0     0  22137      0  0:00:01  0:00:01 --:--:-- 22147


In [9]:
%%sh
ls mydir/

20170228.txt


In [10]:
%%sh
cd mydir/
head 20170228.txt 

Donald J. Trump 

February 28, 2017 
Thank you very much. Mr. Speaker, Mr. Vice President, members of Congress, the first lady of the United States ... 
... and citizens of America, tonight, as we mark the conclusion of our celebration of Black History Month, we are reminded of our nation's path toward civil rights and the work that still remains to be done. 
Recent threats ... 
Recent threats targeting Jewish community centers and vandalism of Jewish cemeteries, as well as last week's shooting in Kansas City, remind us that while we may be a nation divided on policies, we are a country that stands united in condemning hate and evil in all of its very ugly forms. 
Each American generation passes the torch of truth, liberty and justice, in an unbroken chain all the way down to the present. That torch is now in our hands. And we will use it to light up the world. 
I am here tonight to deliver a message of unity and strength, and it is a message deeply delivered from my heart. A new chapt

## To have more fun with all SOU addresses
Do the following:


In [11]:
%%sh
mkdir -p mydir # first create a directory called 'mydir'
cd mydir # change into this mydir directory
rm -f sou.tar.gz # remove any file in mydir called sou.tar.gz
curl -O http://lamastex.org/datasets/public/SOU/sou.tar.gz

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 3566k  100 3566k    0     0  1392k      0  0:00:02  0:00:02 --:--:-- 1392k


In [12]:
%%sh
pwd
ls -lh mydir

/Users/madhurgupta/Documents/HT2024/Introduction to Data Science/Code
total 8312
-rw-r--r--@ 1 madhurgupta  staff    29K Sep  5 08:43 20170228.txt
-rw-r--r--@ 1 madhurgupta  staff   3.5M Sep  5 08:46 sou.tar.gz


In [13]:
%%sh
cd mydir 
tar zxvf sou.tar.gz

x sou/
x sou/18111105.txt
x sou/20040120.txt
x sou/19061203.txt
x sou/18411207.txt
x sou/19091207.txt
x sou/18701205.txt
x sou/19410106.txt
x sou/18571208.txt
x sou/18891203.txt
x sou/18341201.txt
x sou/19660112.txt
x sou/17981208.txt
x sou/19610130.txt
x sou/18140920.txt
x sou/18011208.txt
x sou/18811206.txt
x sou/18281202.txt
x sou/19840125.txt
x sou/18611203.txt
x sou/18731201.txt
x sou/19400103.txt
x sou/19630114.txt
x sou/19281204.txt
x sou/19221208.txt
x sou/19031207.txt
x sou/18681209.txt
x sou/18431206.txt
x sou/18861206.txt
x sou/19261207.txt
x sou/19271206.txt
x sou/19141208.txt
x sou/18791201.txt
x sou/19131202.txt
x sou/19041206.txt
x sou/18001111.txt
x sou/18041108.txt
x sou/20010227.txt
x sou/18621201.txt
x sou/19251208.txt
x sou/19700122.txt
x sou/19790125.txt
x sou/19870127.txt
x sou/20050202.txt
x sou/18331203.txt
x sou/17961207.txt
x sou/18021215.txt
x sou/18771203.txt
x sou/19890209.txt
x sou/18301206.txt
x sou/18121104.txt
x sou/19580109.txt
x sou/20110125.txt
x sou

After running the above two cells, you should have all the SOU (State of Union) addresses. By evaluating the next cell's `ls ...` command you should see the SOU files like the following:

```
total 11M
-rw------- 1 raazesh raazesh 6.6K Feb 18  2016 17900108.txt
-rw------- 1 raazesh raazesh 8.3K Feb 18  2016 17901208.txt
-rw------- 1 raazesh raazesh  14K Feb 18  2016 17911025.txt
...
...
...
-rw------- 1 raazesh raazesh  39K Feb 18  2016 20140128.txt
-rw------- 1 raazesh raazesh  38K Feb 18  2016 20150120.txt
-rw------- 1 raazesh raazesh  31K Feb 18  2016 20160112.txt
```

In [14]:
%%sh
ls -lh mydir/sou

total 21776
-rw-------@ 1 madhurgupta  staff   6.6K Feb 18  2016 17900108.txt
-rw-------@ 1 madhurgupta  staff   8.2K Feb 18  2016 17901208.txt
-rw-------@ 1 madhurgupta  staff    14K Feb 18  2016 17911025.txt
-rw-------@ 1 madhurgupta  staff    12K Feb 18  2016 17921106.txt
-rw-------@ 1 madhurgupta  staff    11K Feb 18  2016 17931203.txt
-rw-------@ 1 madhurgupta  staff    17K Feb 18  2016 17941119.txt
-rw-------@ 1 madhurgupta  staff    12K Feb 18  2016 17951208.txt
-rw-------@ 1 madhurgupta  staff    17K Feb 18  2016 17961207.txt
-rw-------@ 1 madhurgupta  staff    12K Feb 18  2016 17971122.txt
-rw-------@ 1 madhurgupta  staff    13K Feb 18  2016 17981208.txt
-rw-------@ 1 madhurgupta  staff   9.0K Feb 18  2016 17991203.txt
-rw-------@ 1 madhurgupta  staff   8.2K Feb 18  2016 18001111.txt
-rw-------@ 1 madhurgupta  staff    19K Feb 18  2016 18011208.txt
-rw-------@ 1 madhurgupta  staff    13K Feb 18  2016 18021215.txt
-rw-------@ 1 madhurgupta  staff    14K Feb 18  2016 18031017.tx

In [15]:
%%sh
head mydir/sou/17900108.txt

George Washington 

January 8, 1790 
Fellow-Citizens of the Senate and House of Representatives: 
I embrace with great satisfaction the opportunity which now presents itself of congratulating you on the present favorable prospects of our public affairs. The recent accession of the important state of North Carolina to the Constitution of the United States (of which official information has been received), the rising credit and respectability of our country, the general and increasing good will toward the government of the Union, and the concord, peace, and plenty with which we are blessed are circumstances auspicious in an eminent degree to our national prosperity. 
In resuming your consultations for the general good you can not but derive encouragement from the reflection that the measures of the last session have been as satisfactory to your constituents as the novelty and difficulty of the work allowed you to hope. Still further to realize their expectations and to secure the blessin

In [ ]:
%%sh
head mydir/sou/20160112.txt

An interesting analysis of the textual content of the *State of the Union (SoU)* addresses by all US presidents was done in:

-   [Alix Rule, Jean-Philippe Cointet, and Peter S. Bearman, Lexical shifts, substantive changes, and continuity in State of the Union discourse, 1790–2014, PNAS 2015 112 (35) 10837-10844; doi:10.1073/pnas.1512221112](http://www.pnas.org/content/112/35/10837.full).

![](images/F5.large.png)

[Image source](http://www.pnas.org/content/112/35/10837/F5.large.jpg)

[Fig. 5](http://www.pnas.org/content/112/35/10837.full). A river network captures the flow across history of US political discourse, as perceived by contemporaries. Time moves along the x axis. Clusters on semantic networks of 300 most frequent terms for each of 10 historical periods are displayed as vertical bars. Relations between clusters of adjacent periods are indexed by gray flows, whose density reflects their degree of connection. Streams that connect at any point in history may be considered to be part of the same system, indicated with a single color.

**You** *will be able to carry out such analyses and/or critically reflect on the mathematical statistical assumptions made in such analyses, as you learn more during your programme of study after successfully completing this course.*

## How was the `sou.tgz` file created?

If you are curious, read: [http://lamastex.org/datasets/public/SOU/README.md](http://lamastex.org/datasets/public/SOU/README.md).

Briefly, this is how a website with SOU was scraped by Paul Brouwers and adapted by Raazesh Sainudiin. A data scientist, and more generally a researcher interested in making statistical inference from data that is readily available online in a particular format, is expected to be comfortable with such *web-scraping tasks* (which can be done in more gracious and robust ways using specialised Python libraries). Such tasks also known as *Extract-Load-Transform (ELT)* operations are often time-consuming, expensive and the necessary first step towards extracting value from data.

### A bit of bash and lynx to achieve the scraping of the state of the union addresses of the US Presidents,
#### by Paul Brouwers

The code below is mainly there to show how the text content of each state of the union address was scraped from the following URL:

* [http://stateoftheunion.onetwothree.net/texts/index.html](http://stateoftheunion.onetwothree.net/texts/index.html)

Such data acquisition tasks is usually the first and cucial step in a data scientist's workflow.

We have done this and put the data in the distributed file system for easy loading into our notebooks for further analysis.  This keeps us from having to install unix programs like `lynx`, `sed`, etc. that are needed in the shell script below.


```
for i in $(lynx --dump http://stateoftheunion.onetwothree.net/texts/index.html | grep texts | grep -v index | sed 's/.*http/http/') ; do lynx --dump $i | tail -n+13 | head -n-14 | sed 's/^\s\+//' | sed -e ':a;N;$!ba;s/\(.\)\n/\1 /g' -e 's/\n/\n\n/' > $(echo $i | sed 's/.*\([0-9]\{8\}\).*/\1/').txt ; done
```

Or in a more atomic form:

```
for i in $(lynx --dump http://stateoftheunion.onetwothree.net/texts/index.html \

        | grep texts \

        | grep -v index \

        | sed 's/.*http/http/')

do 

        lynx --dump $i \

               | tail -n+13 \

               | head -n-14 \

               | sed 's/^\s\+//' \

               | sed -e ':a;N;$!ba;s/\(.\)\n/\1 /g' -e 's/\n/\n\n/' \

               > $(echo $i | sed 's/.*\([0-9]\{8\}\).*/\1/').txt

done
```

If you have time and are curious how each of the components in the above pipeline via `|` operators work, try to read `man echo`,  `man sed`,  `man grep`,  `man head`,  `man tail`, and  `man lynx` or `lynx --help`. If a command like `lynx` is not in your system, then you can install it with some work (mostly googling).

In [16]:
%%sh
## uncomment by removing '#' in the next line and try executing this cell by pressing Ctrl-Enter to see if lynx is installed
lynx --help

USAGE: lynx [options] [file]
Options are:
  -                 receive options and arguments from stdin
  -accept_all_cookies 
                    accept cookies without prompting if Set-Cookie handling
                    is on (off)
  -anonymous        apply restrictions for anonymous account,
                    see also -restrictions
  -assume_charset=MIMEname
                    charset for documents that don't specify it
  -assume_local_charset=MIMEname
                    charset assumed for local files
  -assume_unrec_charset=MIMEname
                    use this instead of unrecognized charsets
  -auth=id:pw       authentication information for protected documents
  -base             prepend a request URL comment and BASE tag to text/html
                    outputs for -source dumps
  -bibhost=URL      local bibp server (default http://bibhost/)
  -book             use the bookmark page as the startfile (off)
  -buried_news      toggles scanning of news articles for buried ref

So using `lynx` is not that difficult. Suppose you want to dump the contents of [https://lamastex.github.io/research/#available-student-projects](https://lamastex.github.io/research/#available-student-projects) to `stdout` or standard out, we can do the following:

In [17]:
%%sh
## uncomment by removing '#' in the next line and try executing this cell if lynx is installed
lynx --dump https://lamastex.github.io/research/#available-student-projects

   #[1]LaMaStEx Feed

   [2]LaMaStEx
     * [3]About
     * [4]Contact
     * [5]Research
     * [6]CV
     * [7]Publications
     * [8]Software
     * [9]Courses
     * [10]Talks
     * [11]Students

   (BUTTON)

Research and Available Student Projects

   At the interface of computing, mathematics and statistics.

   Raazesh Sainudiin

Raazesh Sainudiin

   I work at the interface of computing, mathematics and statistics to
   solve real-world problems through custom-built mathematical and
   statistical models.
   (BUTTON) Follow
     * Uppsala, Sweden
     * Email
     * [12]LinkedIn
     * [13]GitHub
     * [14]YouTube

   [ ] Toggle Menu
     * [15]LaMaStEx
          + [16]About
          + [17]Contact
     * [18]Research
          + [19]CV
          + [20]Publications
          + [21]Software
          + [22]Talks
          + [23]Students
     * [24]Courses
          + [25]Scalable Data Science

On This Page

     * [26]Research Overview
     * [27]Available Student Projects
   

Hopefully, you had fun with BASH! Now let us put BASH to use.